feat: integrate agentAPI with resources monitoring logic

defelmnq opened the pull request:
As part of the new resources monitoring logic - more specifically for OOM & OOD Notifications , we need to update the AgentAPI , and the agents logic.

This PR aims to do it, and more specifically :  
We are updating the AgentAPI & TailnetAPI to version 24 to add two new methods in the AgentAPI : 
- One method to fetch the resources monitoring configuration
- One method to push the datapoints for the resources monitoring.

Also, this PR adds a new logic on the agent side, with a routine running and ticking - fetching the resources usage each time , but also storing it in a FIFO like queue.

Finally, this PR fixes a problem we had with RBAC logic on the resources monitoring model, applying the same logic than we have for similar entities.

defelmnq committed:
work on new agent version

defelmnq committed:
improve function for resources monitoring

defelmnq committed:
add missing files

defelmnq committed:
work on resources monitor tests

defelmnq committed:
apply fmt and lint

defelmnq committed:
work on dbauthz tests

defelmnq committed:
work on dbauthz

defelmnq committed:
work on rbac

defelmnq committed:
continue to iterate

defelmnq committed:
continue to iterate

defelmnq committed:
continue to iterate

defelmnq committed:
work on tests

defelmnq committed:
improve testing

defelmnq commented on agent/proto/agent.proto:
```diff
@@ -295,6 +295,42 @@ message Timing {
     Status status = 6;
 }
 
+message GetResourcesMonitoringConfigurationRequest {
```
This message is based on what has been described in the RFC - and has been mutually discussed and approved by Vincent and @DanielleMaywood

defelmnq commented on agent/resources_monitor.go:
```diff
@@ -0,0 +1,135 @@
+	"github.com/coder/quartz"
+)
+
+func (a *agent) pushResourcesMonitoring(ctx context.Context, aAPI proto.DRPCAgentClient24) error {
```
self review : We have this intermediate function in order to have an extra layer, and simplicify for testing purpose. 
As all arguments are injected as parameters to `PushResourcesMonitoringWithConfig` we can use the mocked versions.

defelmnq commented on agent/resources_monitor.go:
```diff
@@ -0,0 +1,135 @@
+	configFetcher ResourcesMonitorConfigurationFetcher,
+	datapointsPusher ResourcesMonitorDatapointsPusher,
+) error {
+	config, err := configFetcher(ctx, &proto.GetResourcesMonitoringConfigurationRequest{})
```
self review : There's three situations on the which we do not want to run the go-routine : 
- If we can not fetch the configuration from Coderd.
- If the resources_monitoring is not enabled.
- If we can not instantiate the clistat client (which is the one used to fetch resources usage.)

defelmnq commented on coderd/agentapi/resources_monitoring.go:
```diff
@@ -0,0 +1,55 @@
+	}, nil
+}
+
+func (a *ResourcesMonitoringAPI) PushResourcesMonitoringUsage(ctx context.Context, req *proto.PushResourcesMonitoringUsageRequest) (*proto.PushResourcesMonitoringUsageResponse, error) {
```
self review : This function is just here as a placeholder - we are fine keeping it this way for now as it will be replaced by @DanielleMaywood PR.

defelmnq commented on agent/resources_monitor.go:
```diff
@@ -0,0 +1,135 @@
+
+	datapointsQueue := NewResourcesMonitorQueue(int(config.Config.NumDatapoints))
+
+	clk.TickerFunc(ctx, time.Duration(config.Config.TickInterval*int32(time.Second)), func() error {
```
self review : We prefer to log instead of returning an error in case of problem.

Using `TickerFunc` from quartz, returning an error would just shutdown the whole ticker and stop it. Returning nil stops the current tick, but the next one will still trigger.

defelmnq committed:
improve error messages

defelmnq commented:
@spikecurtis I'd love to have your review - specially on the whole AgentAPI / TailnetAPI part - I want to be sure that I increase the version as we should.

dannykopping commented on agent/resources_monitor.go:
```diff
@@ -0,0 +1,132 @@
+
+	datapointsQueue := NewResourcesMonitorQueue(int(config.Config.NumDatapoints))
+
+	clk.TickerFunc(ctx, time.Duration(config.Config.TickInterval*int32(time.Second)), func() error {
```
I believe you need to call `Wait()` on the resulting value.

defelmnq replied:
From my understanding of the quartz lib , `Wait()` is called when the TickerFunc is globally finished (either the associated context expires, or one of the ticks returns an error) - and block until this moment. 

I designed this function to just start the ticker and then return , using the context to control it. 

I can change it if you see some added value, but am not sure.

dannykopping commented on agent/resources_monitor.go:
```diff
@@ -0,0 +1,132 @@
+	ResourcesMonitorDatapointsPusher     func(ctx context.Context, params *proto.PushResourcesMonitoringUsageRequest) (*proto.PushResourcesMonitoringUsageResponse, error)
+)
+
+func PushResourcesMonitoringWithConfig(ctx context.Context,
```
This implementation does not seem to match the RFC design.

Once the buffer fills up, it'll send the full buffer to the control plane. However, it'll keep doing so every `TickInterval` thereafter since the buffer only ever replaces the oldest value.

The RFC specifies that the payload should only be delivered every 15s for the 20 datapoints; it's currently performing collection and submission in the same ticker with no distinction between them.

We also seem to not be adding the `UNKNOWN` datapoints when values cannot be retrieved.

Was this a conscious choice? Is the RFC out-of-date perhaps?

defelmnq replied:
About the `UNKNOWN` status - this is handled directly on the coderd side. 
It will be visible in the part currently being done by Danielle - from the configuration stored in the DB, and what we return on the Agent side, coderd (and the processing component) decide if the resource, for the given datapoint, is `OK`, `NOK` or `UNKNOWN`. 

About the tick logic - that's maybe a misunderstanding in the RFC , but when I defined it to be delivered every 15s, it was based on the `TickInterval` - considering it was set to 15. having both the `TickInterval` and another value set independently would require some sync to avoid sending two times the datapoints without having a refresh or fetching datapoints too frequently and so not pushing data to coderd at each tick. 

TL;DR - I think the best solution is to have the `TickInterval` handling both the refresh of resources and push to coderd so we are sure each push will always have the last data.

dannykopping replied:
> About the UNKNOWN status - this is handled directly on the coderd side.
It will be visible in the part currently being done by Danielle - from the configuration stored in the DB, and what we return on the Agent side, coderd (and the processing component) decide if the resource, for the given datapoint, is OK, NOK or UNKNOWN.

Can you point me to some specific code I can look at please? Will you be inferring that a datapoint was missed by looking at the time of each datapoint? I wonder why we opted for this approach rather than explicitly sending a noop datapoint to indicate a problem.

> TL;DR - I think the best solution is to have the TickInterval handling both the refresh of resources and push to coderd so we are sure each push will always have the last data.

This doesn't really address what I'm asking, though.

We're going to be sending all _n_ datapoints every `TickInterval` once the queue fills up since we never drain the queue but only shifting off the oldest element.

dannykopping reviewed (commented):
Generally looking good, but I'm concerned that this doesn't match the RFC. See my comment in `agent/resources_monitor.go`.

spikecurtis reviewed (changes requested):
Proto version changes look overall fine, but I want to review again re: RBAC

defelmnq commented on agent/proto/agent.proto:
```diff
@@ -295,6 +295,42 @@ message Timing {
+	bool enabled = 1;
+	message Config {
+		int32 num_datapoints = 1;
+		int32 tick_interval = 2; // expressed in seconds
```
I like and agree with the proposal - changed ✅

defelmnq committed:
rework architecture of resources monitor

defelmnq commented on agent/resources_monitor_queue_test.go:
```diff
@@ -0,0 +1,102 @@
+			}
+			if got := queue.Items(); !equal(got, tt.expected) {
+				t.Errorf("expected %v, got %v", tt.expected, got)
+			}
```
Mostly due to how these tests started when I was writting it - but fixed to use require and improve a bit the logic too ✅

defelmnq commented on coderd/agentapi/resources_monitoring.go:
```diff
@@ -0,0 +1,55 @@
+	}
+
+	return &proto.GetResourcesMonitoringConfigurationResponse{
+		Enabled: false,
```
You're right -

To clarify, this boolean defines if the monitoring is enabled or not - which means are we starting the ticker that will fetch the resources and push it back to coderd. 

As the agent PR is part of a bigger topic, including another PR worked on by Danielle - for now monitoring should be turned off. 

Once everything merged and we're ready to deploy, we'll just have to turn this one on with the release of coder/coder.

defelmnq commented on agent/proto/agent.proto:
```diff
@@ -295,6 +295,42 @@ message Timing {
+}
+
+message GetResourcesMonitoringConfigurationResponse {
+	bool enabled = 1;
```
The two options should be possible - We should be able to enable/disable the resources monitoring globally, but also do it for a given volume or memory.

The one appearing here is the boolean about the overall resources monitoring (I currently add the other one.) - do you have an idea of what you'd like to change ?

spikecurtis replied:
At the protocol level, I don't think it makes sense to have a global enable/disable switch in addition to enable/disable for individual components. A global disable is equivalent to disabling each individual thing, no? If so then it is redundant and can lead to confusion. Let's just stick to enable/disable for memory and individual volumes.

defelmnq replied:
i changed the payload for a version matching more what you described : 
- Removed the 'global' enabled boolean.
- Added a boolean (which was missing) on each resource
- Added a field for the memory

This new version seems pretty clean to me :)

defelmnq commented on agent/resources_monitor.go:
```diff
@@ -0,0 +1,132 @@
+
+type (
+	ResourcesMonitorConfigurationFetcher func(ctx context.Context, params *proto.GetResourcesMonitoringConfigurationRequest) (*proto.GetResourcesMonitoringConfigurationResponse, error)
+	ResourcesMonitorDatapointsPusher     func(ctx context.Context, params *proto.PushResourcesMonitoringUsageRequest) (*proto.PushResourcesMonitoringUsageResponse, error)
```
I removed the `func` parts and `ResourcesMonitorConfigurationFetcher` to have an interface - should be way better this way. 👀

defelmnq commented on agent/resources_monitor.go:
```diff
@@ -0,0 +1,132 @@
+	datapointsPusher ResourcesMonitorDatapointsPusher,
+) error {
+	config, err := configFetcher(ctx, &proto.GetResourcesMonitoringConfigurationRequest{})
+	if err != nil {
```
✅ changed too.

defelmnq commented on agent/resources_monitor.go:
```diff
@@ -0,0 +1,132 @@
+package agent
```
I moved it to its own package ✅ should indeed simplify a lot the naming.

defelmnq commented on agent/resources_monitor.go:
```diff
@@ -0,0 +1,132 @@
+		return xerrors.Errorf("failed to create resources fetcher: %w", err)
+	}
+
+	datapointsQueue := NewResourcesMonitorQueue(int(config.Config.NumDatapoints))
```
I changed a lot the way things were organised based on the comments - so now : 
- Resources monitor logic is in its own package
- It works with a struct and methods to make it an entity
- members are used to store the queue & config

🙏

DanielleMaywood commented on agent/proto/resourcesmonitor/resources_monitor.go:
```diff
@@ -0,0 +1,130 @@
+
+	datapointsQueue := NewQueue(int(m.config.Config.NumDatapoints))
+
+	m.clock.TickerFunc(ctx, time.Duration(m.config.Config.CollectionIntervalSeconds*int32(time.Second)), func() error {
```
I don't think this duration is doing what we'd expect.

```suggestion
	m.clock.TickerFunc(ctx, time.Duration(m.config.Config.CollectionIntervalSeconds)*time.Second, func() error {
```

I've put together a go playground demonstrating the issue.

https://go.dev/play/p/IDW5tcb5FWS

defelmnq replied:
Nice catch here ! Changed 🙈

defelmnq committed:
improve resourcesmonitor struct

defelmnq committed:
improve resourcesmonitor struct

DanielleMaywood commented on coderd/agentapi/resources_monitoring.go:
```diff
@@ -0,0 +1,57 @@
+}
+
+func (a *ResourcesMonitoringAPI) GetResourcesMonitoringConfiguration(ctx context.Context, _ *proto.GetResourcesMonitoringConfigurationRequest) (*proto.GetResourcesMonitoringConfigurationResponse, error) {
+	agent, err := a.AgentFn(ctx)
```
nit:

I think we could avoid using the `AgentFn` pattern here. We only need the Agent's ID, but this function makes a call to the database, using the Agent's ID, to fetch the agent. We could instead have the `ResourcesMonitoringAPI` have an `AgentID` field and just pass in the `AgentID` from `Options`.

defelmnq replied:
A bit surprised there's a so simple option , but makes sense and changed - Thanks ! ✅

defelmnq committed:
change proto payload for get resources monitoring config

defelmnq committed:
change proto payload for get resources monitoring config

defelmnq commented on coderd/agentapi/resources_monitoring.go:
```diff
@@ -0,0 +1,55 @@
+	}
+
+	_, err = a.Database.FetchMemoryResourceMonitorsByAgentID(ctx, agent.ID)
+	if err != nil && !errors.Is(err, sql.ErrNoRows) {
```
Yep - I indeed skipped this one on the first proto version which I just changed. New payload seem way better.

- Added an enabled field for each resource (each volume and the memory)
- Removed the enabled field from global configuration ([more info here](https://github.com/coder/coder/pull/16438#discussion_r1950355950))
- Memory now is optional

defelmnq commented on agent/proto/agent.proto:
```diff
@@ -295,6 +295,42 @@ message Timing {
+
+		google.protobuf.Timestamp collected_at = 1;
+		MemoryUsage memory = 2;
+		repeated VolumeUsage volumes = 3;
```
I changed the `memory` to be optional but it was (and still it) a pointer.

For me, purely in Go, it means we can verify pretty easily if memory is nil or not - and just iterate over the slice of volumes to get all the resources.

If you have another idea, happy to discuss it.

defelmnq commented on agent/resources_monitor_test.go:
```diff
@@ -0,0 +1,121 @@
+			expectedError: false,
+			numTicks:      60,
+			counterCalls:  0,
+			expectedCalls: 41,
```
I removed the `expectedCalls` from the struct to be computed on the test function directly - trying to remove the magic.

Please tell me if that helps.

defelmnq committed:
rework fetcher and tests

defelmnq commented on agent/resources_monitor_test.go:
```diff
@@ -0,0 +1,121 @@
+func TestPushResourcesMonitoringWithConfig(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
```
I improved a lot the testability of this block by creating a struct for the resources_fetcher part - so we can mock it.

Also added test cases for all the missing cases. Should be way better now.

defelmnq commented:
@dannykopping @spikecurtis thanks for the first review. I changed and iterated on most of the things excepted RBAC logic as it will be a whole topic and I dont want to block the rest.

Can I ask you for another review (without RBAC) - ensuring we are all good for the whole logic ? 

Globally : 
- I moved the ResourcesMonitor part to its own package.
- Created a fetcher interface to improve testability (and added a few cases)
- Changed the proto part to have optional fields , and remove the global `enabled` in favor of enabled for each resource.
- Some cleanup here and there based on comments

Thanks !

defelmnq committed:
fix tests

defelmnq committed:
fix tests

defelmnq committed:
fix tests

defelmnq committed:
fix logic

spikecurtis commented:
I'm happy to give another review, but I consider the RBAC to be a blocker to merging.

spikecurtis commented on agent/proto/resourcesmonitor/fetcher.go:
```diff
@@ -0,0 +1,61 @@
+	"github.com/coder/coder/v2/cli/clistat"
+)
+
+type ResourcesFetcher interface {
```
this can just be called `Fetcher` with methods `FetchMemory` and `FetchVolume` -- the package name is resourcesmonitor, and Go style is to reduce "stutter" as much as possible.

spikecurtis commented on agent/proto/resourcesmonitor/fetcher.go:
```diff
@@ -0,0 +1,61 @@
+}
+
+type resourcesFetcher struct {
+	fetcher *clistat.Statter
```
```suggestion
	*clistat.Statter
```

Creating a field called "fetcher" in a struct called "fetcher" is confusing. You can just make the struct just contain the Statter and add additonal methods.

spikecurtis commented on agent/proto/resourcesmonitor/fetcher_test.go:
```diff
@@ -0,0 +1,14 @@
+
+type resourcesFetcherMock struct {
+	fetchResourcesMonitoredMemoryFunc func() (total int64, used int64, err error)
+	fetchResourcesMonitoredVolumeFunc func(volume string) (total int64, used int64, err error)
```
In your tests you return static values for the monitored resources, so I think it's much clearer to define a struct with the static values and then have that struct implement the interface, rather than create this extra layer of indirection.

spikecurtis commented on agent/proto/resourcesmonitor/fetcher.go:
```diff
@@ -0,0 +1,61 @@
+	return volTotal, volUsed, nil
+}
+
+func (*resourcesFetcher) bytesToMegabytes(bytes int64) int64 {
```
Why are we converting to megabytes? An int64 is plenty big to hold up to exabytes, so it seems like we are dropping precision for no good reason.

There is no indication in the `.proto` that these should be measured in MiB, so if we are going to use any unit other than byte, it should be in the name.

defelmnq replied:
Yeah it was an arbitrary decision done as the first version was done with int32 in proto. Now that we switched everything at int64 , bytes makes sense and give us more flexibility. 
Switched :)

defelmnq committed:
improve testing fetcher and rename struct

defelmnq committed:
lint

defelmnq committed:
work on dbauthz

defelmnq committed:
improve dbauthz for fetching

defelmnq committed:
change dbauthz permissions

defelmnq committed:
finalise tests

defelmnq committed:
fix comments from github

defelmnq commented:
Thanks @spikecurtis for the other review. I think we reached something way better here : 

- I fixed your last comments
- re-inserted the RBAC logic. Took a bit of time understanding the way everything works but now tested and am happy with the result.

Provisionerd can add entries
Whoever can read the associated workspace can also fetch the resources_monitoring associated.

dannykopping commented on agent/proto/resourcesmonitor/resources_monitor_test.go:
```diff
@@ -0,0 +1,235 @@
+			// We have one call per tick, once reached the ${config.NumDatapoints}.
+			expectedCalls := tt.numTicks - int(tt.config.Config.NumDatapoints) + 1
+			require.Equal(t, expectedCalls, counterCalls)
+			cancel()
```
Nit: I don't think you need this here since you have a `defer cancel()` already which will be invoked after this.

dannykopping reviewed (changes requested):
I still have concerns about how this data collection is implemented.

`GetResourcesMonitoringConfiguration` specifies that a collection will occur every 10 seconds. Datapoints will be added to the queue, and once the queue fills up it'll send the contents of the queue to `coderd`. Once the queue is full it never empties, so it'll just shift the oldest element off to accommodate a new one.

The `CollectionIntervalSeconds` defines both the collection AND send interval, which means that once the queue is full it'll send the full contents every `CollectionIntervalSeconds`, but...

`CollectionIntervalSeconds` is 10s by default, which is too slow for datapoint collection; the point of the system we discussed in the RFC was to collect _often_ (maybe even once per second), and to buffer those datapoints to send every so often so as to not overwhelm `coderd`. This implementation means we're sending the _same_ datapoint `NumDatapoints-1` times to `coderd`. I don't see the point in this.

Regarding the processor: with this design it would take up to 200s / 3.3m (20 datapoints collected 10s apart) to go from an unhealthy state to a healthy state; that will not feel responsive, and we can do much better!

~In [another comment](https://github.com/coder/coder/pull/16438#discussion_r1950351012) I described my reservations about this design concerning failed collections; from the answer it seems like `coderd` will be _inferring_ those failed collections by noticing gaps in the data; how we'll get this precise I don't know (since we're not collecting timestamps with each datapoint) - and it seems unnecessary since we already know at the source when this has occurred. Currently we're just skipping over these failed datapoints in the agent with a log, and leaving `coderd` to guess as to what happened; this doesn't seem like a resilient design to me.~

@defelmnq pointed out that an empty datapoint would be sent if the collection failed (I missed this on line 73). `coderd` will also look at the monitoring config and determine if the datapoint is blank but shouldn't be, and determine that as `UNKNOWN`. I think it could've been more explicit from the agent's perspective, but this approach will work fine.

spikecurtis reviewed (approved):
RBAC stuff looks OK now.

dannykopping reviewed (approved):
OK, @defelmnq and I had discussed the implementation and I'm happy to unblock this as it stands.
What I thought we were building and what I agreed to in the RFC were different, and for that I take responsibility. Apologies for the inertia this caused on this review!

One optimisation we agreed to do now is to start shipping the payloads even before the buffer fills up, so we can alert early if errors are present straight away.

---

For later:

My argument is that we should collect faster and send at the ~same rate, to achieve a higher resolution measurement and consequently have more reactive alerts / status updates.

To achieve this, we would need to send at least half (but probably the full) previous `NumDatapoints` datapoints to ensure we can satisfy the 3 rules (which may only be true across current + previous payloads):

- 4 datapoints in a row above threshold = `NOK`
- any 10 datapoints in a `NumDatapoints` payload above threshold = `NOK`
- all `NumDatapoints` within threshold = `OK`

defelmnq committed:
add collectedAt
